bson/codec_options.py

```diff
@@ -126,9 +160,14 @@ def __new__(cls, document_class=dict,
                 raise ValueError(
                     "cannot specify tzinfo without also setting tz_aware=True")
 
+        type_registry = type_registry or TypeRegistry()
```

Comment:
I don't think we can have the `or TypeRegistry()` here. Creating a new TypeRegistry for each CodecOptions breaks equality checks. We should leave it as `None` and check for `None` before calling any methods on `opts.type_registry`.

Reply:
How about implementing `__eq__` for `TypeRegistry` instead? I imagine equality checks on `CodecOptions` instances will be quite rare so performance isn't a major consideration (since comparing `None` values will obviously be faster than comparing 2 dictionaries). Also, I really don't envy the thought of sanity-checking `type_registry` before using it everywhere in our code. 
```
def __eq__(self, other):
    return (self._encoder_map == other._encoder_map) and (self._decoder_map == other._decoder_map)
```

Reply:
STGM!